### ISSUE TITLE
Fix FN S4325 (`no-unnecessary-type-assertion`): flags necessary not-null assertion as unnecessary

### ISSUE THREAD

**Issue by thw0rted** [SonarSource/SonarJS#1961]
**I want to report a bug.**

**SonarJS version:** 6.2
**SonarQube version:** 6.7.6

**Rule key:** typescript:S4325

**Reproducer** 
````ts
interface Foo {
  bar: { baz: number } | undefined;
}

const foo: Foo = {
  bar: { baz: 1 }
};

const bar = foo.bar!;

console.log(bar.baz);
````

**Expected behavior**

The not-null assertion is required to compile cleanly with `strictNullChecks` and the rule should not flag it. 

With the assertion, the type of `bar` is `{baz: number}`. Without it, the type is `{baz: number} | undefined`.  This is why, with `strictNullChecks`, the compiler would error on the last line with "Object is possibly undefined".

**Comment by thw0rted** [comment]
Maybe the problem is larger than this.  I also get a false positive when passing a nullable type to a function that requires a non-null value, e.g.

````ts
interface Foo {
  bar?: string;
}

function f(s: string){ }

const foo: Foo = { bar: "hey" };
f(foo.bar!);
````

Again, in strict mode, the compiler would flag the last line if I removed the assertion but Sonar marks it as "unnecessary".

**Comment by thw0rted** [comment]
I found another one already! Same rule, maybe the same conflict with `strictNullChecks`?

````ts
interface Foo{
  bar: string;
}

function f(f: Foo){ }

const foo: Partial<Foo> = {};
foo.bar = "hey";
f(foo as Foo);
````

If I don't cast on that last line, the compiler flags 

````text
Argument of type 'Partial<Foo>' is not assignable to parameter of type 'Foo'.
  Types of property 'bar' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.ts(2345)
````

**Comment by saberduck** [comment]
hello @thw0rted , what is the version of TypeScript you use?

**Comment by thw0rted** [comment]
I typically keep up with current releases, at least within a month or so.  I'm currently using 4.0.3.  Does that have an impact on this issue?  Do you think some of these are only caught in later versions of TS?

**Comment by saberduck** [comment]
Usually, a more recent version of TS is smarter, so some not-null assertions are not needed in later versions. But 4.0.3 is very recent, so it shouldn't be case here.

**Comment by Naxos84** [comment]
I'm using still TS 3.8.3 and have a the same issue in the following case.

```
const result = someString.split("?").shift()!;
```

I know that there could be no "?" inside the string and the `shift` would return `undefined`.
But even if I make sure that the string contains a "?" it is complaining about a unnecessary cast.

**Comment by thw0rted** [comment]
`String#split` returns a `string[]`, and `Array#shift()` returns `string | undefined`.  (There has been talk on the TS team, especially since https://github.com/microsoft/TypeScript/pull/39560 got implemented, about making `split` return a tuple (like `[string, ...string[]]`) to guarantee that the first element is not undefined, but apparently there is an edge case where `"".split("")` returns an empty array so that isn't possible.)

At any rate, the practical upshot is that your assertion is "necessary" but only because you know that `split("not an empty string")` cannot return a zero-element array, and you don't want your variable typed as `| undefined`.  Does the rule still trigger if you explicitly type the variable as `string`?

**Comment by Naxos84** [comment]
I checked it with the following code.

```
function sonarLintDemo(value: string): string {
    if (!value.includes("?")) {
        return "";
    }
    const firstArrayElement = value.split("?").shift()!; //Sonarlint is complaining here.

    return firstArrayElement;
}
```

**Comment by thw0rted** [comment]
Sorry, to clarify, I meant, if it complains about

````ts
    const firstArrayElement = value.split("?").shift()!;
````

maybe it wouldn't complain about

````ts
    const firstArrayElement: string = value.split("?").shift()!;
````

because one could argue that the former is not "necessary" -- you're allowing the compiler to infer type type of `firstArrayElement` -- but the latter is, *if* you want the r-value type to exactly match the l-value.

**Comment by Naxos84** [comment]
Nope.
It just changes location of error message.

from `return firstArrayElement;`
to `const firstArrayELement: string = value .....`

**Comment by cpiber** [comment]
Related: typescript:S4325 flags casts to more specific interfaces as unnecessary, e.g. from `HTMLElement` to `HTMLInputElement`.

**Comment by shubham-shewale** [comment]
Is there any update on this ?
SonarLint if still flagging an issue when I am typecasting from `Element` to `HTMLElement.`
```
const selector = (<HTMLElement>doc.querySelector('#spnValue'));
let curText = selector.innerText.trim();
```

**Comment by cpiber** [comment]
@shubham-shewale For querySelector I would recommend using the proper instantiation of the generic function instead of casting the result:
```ts
const selector = doc.querySelector<HTMLElement>('#spnValue');
```
(just my two cents)

**Comment by shubham-shewale** [comment]
@cpiber Indeed this is very helpful !! 
Thanks a lot !!

**Comment by cpiber** [comment]
@shubham-shewale Do you mean generics in general? These are explained here: https://www.typescriptlang.org/docs/handbook/2/generics.html

**Comment by shubham-shewale** [comment]
> Related: typescript:[S4325](https://sonarsource.github.io/rspec/#/rspec/S4325/javascript) flags casts to more specific interfaces as unnecessary, e.g. from `HTMLElement` to `HTMLInputElement`.

@cpiber I was looking for the document/Explanation about the mentioned scenario `casts to more specific interfaces as unnecessary`. ( Just out of curiosity wanted to know more why such behavior ).

Thanks for sharing the Documentation on Generic !!

**Comment by MichaelMHoff** [comment]
I think we are experiencing the same issue:

VS Code:
![image](https://github.com/SonarSource/SonarJS/assets/125384031/84e140b5-938f-49ff-9aea-e54adcde6c00)
![image](https://github.com/SonarSource/SonarJS/assets/125384031/e460f543-a6fa-467e-ad1c-4a59bedc4417)

Sonarcloud:
<img width="685" alt="image" src="https://github.com/SonarSource/SonarJS/assets/125384031/1810570d-4036-4dda-b289-840b0eb13675">

strictNullChecks = true
typescript = 4.9.4

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-138
